src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs

```diff
@@ -26,44 +26,110 @@ namespace OpenTelemetry.Instrumentation.Http.Implementation
         {
             if (name == OnStopEvent)
             {
-                if (Sdk.SuppressInstrumentation)
-                {
-                    return;
-                }
+                this.OnStopActivity(Activity.Current, payload);
+            }
+        }
 
-                var activity = Activity.Current;
-                if (this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) && response != null)
-                {
-                    var request = response.RequestMessage;
+        private void OnStopActivity(Activity activity, object payload)
+        {
+            if (Sdk.SuppressInstrumentation)
+            {
+                return;
+            }
 
-                    TagList tags = default;
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpMethod, HttpTagHelper.GetNameForHttpMethod(request.Method)));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpScheme, request.RequestUri.Scheme));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpStatusCode, (int)response.StatusCode));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpFlavor, HttpTagHelper.GetFlavorTagValueFromProtocolVersion(request.Version)));
-                    tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeNetPeerName, request.RequestUri.Host));
+            if (!this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) || response == null)
+            {
+                // TODO: logging?
+                return;
+            }
+
+            var request = response.RequestMessage;
+            if (!this.TryFilterHttpRequestMessage(activity.OperationName, request))
+            {
+                return;
+            }
+
+            TagList tags = default;
+            tags.Add(new(SemanticConventions.AttributeHttpMethod, HttpTagHelper.GetNameForHttpMethod(request.Method)));
+            tags.Add(new(SemanticConventions.AttributeHttpScheme, request.RequestUri.Scheme));
+            tags.Add(new(SemanticConventions.AttributeHttpStatusCode, (int)response.StatusCode));
+            tags.Add(new(SemanticConventions.AttributeHttpFlavor, HttpTagHelper.GetFlavorTagValueFromProtocolVersion(request.Version)));
+            tags.Add(new(SemanticConventions.AttributeNetPeerName, request.RequestUri.Host));
+
+            if (!request.RequestUri.IsDefaultPort)
+            {
+                tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeNetPeerPort, request.RequestUri.Port));
+            }
 
-                    if (!request.RequestUri.IsDefaultPort)
-                    {
-                        tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeNetPeerPort, request.RequestUri.Port));
-                    }
+            this.EnrichWithHttpRequestMessage(ref tags, request);
```

Comment:
same as above...this looks like response message and not request?

Reply:
In meter, both request and response are used. Should I do the same? I agree that it's more convinient to have response, but it'll conflict with https://github.com/open-telemetry/opentelemetry-dotnet/pull/4098